### PR TITLE
Refactored and added tests to verify sending and receiving of all byte values

### PR DESCRIPTION
**Pull request by bainian12345**
Refactored and added tests to verify sending and receiving of all byte values

### CODE REVIEW

File: src/main/java/com/microsoft/azure/relay/ClientWebSocket.java [hunk 1]
```diff
@@ -247,6 +247,7 @@ else if (data instanceof String) {
 							}
 							
 							int bytesToSend = bytes.remaining();
+							// sendBinary() will cause the content of the byte array within the ByteBuffer to change
```
Comment (dlstucki) [inline src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: >content of the byte array [](start = 38, length = 25)

This is weird. What do the array contents become?

Comment (bainian12345) [reply src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: Becomes garbage values...I haven't looked into it, not sure what they are doing with the array after sending

---
In reply to: [294013165](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294013165) [](ancestors = 294013165)

Comment (dlstucki) [reply src/main/java/com/microsoft/azure/relay/ClientWebSocket.java]: no reason, I'm just surprised they party on someone else's buffer

---
In reply to: [294013165](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294013165) [](ancestors = 294013165)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 2]
```diff
@@ -326,18 +392,28 @@ private static void sendResponseMessages(RelayedHttpListenerContext context, Str
...
+				break;
+			} else {
+				byteStream.write(i);
 			}
 		}
-		return builder.toString();
+		return byteStream.toByteArray();
+	}
+	
+	private byte[] getByteSet() {
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >getByteSet [](start = 16, length = 10)

Do all these byte-level tests use exactly 256bytes?

Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: All 256 bytes are checked for equality

---
In reply to: [294017638](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294017638) [](ancestors = 294017638)

Comment (dlstucki) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: At some point it would be nice to check that an odd number of bytes work. 1, 255, 1024, 4k, 64k-1, 64k, 64k+1, etc. In the future consider having getByteSet take a size and fill an array of that size with this repeating byte pattern.

---
In reply to: [294017638](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294017638) [](ancestors = 294017638)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 3]
```diff
@@ -75,29 +76,33 @@ public void createClient() throws URISyntaxException {
...
+		CompletableFuture<Void> clientTask = websocketClient(largeStr, smallStr);
 		listenerTask.join();
+		clientTask.join();
 	}
 	
 	@Test
 	public void websocketLargeSendSmallResponseTest() {
 		CompletableFuture<Void> listenerTask = websocketListener(largeStr, smallStr);
-		websocketClient(smallStr, largeStr);
+		CompletableFuture<Void> clientTask = websocketClient(smallStr, largeStr);
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >websocketClient [](start = 39, length = 15)

nothing to do in this PR, I just want to point out this method name is a noun only and method names are more self-documenting when they contain verbs too.

Comment (dlstucki) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: e.g. RunClientTest, TestClient, ProcessClient, ConfigureClient, etc.

---
In reply to: [294378342](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294378342) [](ancestors = 294378342)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 4]
```diff
@@ -170,6 +175,34 @@ public void websocketMultipleSenderTest() throws URISyntaxException {
...
+					byte[] byteSet = getByteSet();
+					assertTrue("Websocket listener did not receive the expected bytes.", Arrays.equals(byteSet, bufferReceived.array()));
+					return connection.writeAsync(ByteBuffer.wrap(byteSet));
+				})
+				.thenCompose($void -> {
+					return connection.closeAsync(new CloseReason(CloseCodes.NORMAL_CLOSURE, "Normal closure from listener"));
+				});
+		});
+		
+		HybridConnectionClient newClient = new HybridConnectionClient(new URI(TestUtil.RELAY_NAMESPACE_URI + TestUtil.ENTITY_PATH), tokenProvider);
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >newClient [](start = 25, length = 9)

should this test ideally close the HCClient no matter what?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 5]
```diff
@@ -240,6 +273,32 @@ public void httpWriteResponseEnsureOrderTest() throws IOException {
 		httpRequestSender("POST", String.join("", messages), smallStr);
 	}
 	
+	@Test
+	public void httpWriteReceiveAllByteValuesTest() throws IOException {
+		byte[] byteSet = getByteSet();
+		listener.setRequestHandler(context -> {
+			RelayedHttpListenerResponse response = context.getResponse();
+			response.setStatusCode(STATUS_CODE);
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >setStatusCode [](start = 12, length = 13)

Are there any tests in this suite which use any other status code?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 6]
```diff
@@ -240,6 +273,32 @@ public void httpWriteResponseEnsureOrderTest() throws IOException {
...
+			response.setStatusDescription(STATUS_DESCRIPTION);
+			
+			try {
+				response.getOutputStream().writeAsync(byteSet, 0, byteSet.length).join();
+			} catch (Exception e) {
+				fail(e.getMessage());
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >fail [](start = 4, length = 4)

If this fail occurs in the requestHandler callback does the unit test fail? (Actually, I think so because the correct response most likely won't appear.

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 7]
```diff
@@ -240,6 +273,32 @@ public void httpWriteResponseEnsureOrderTest() throws IOException {
...
+			} finally {
+			    context.getResponse().close();
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >context.getResponse() [](start = 7, length = 21)

could this be "response.close()"?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 8]
```diff
@@ -280,29 +339,36 @@ private static void httpRequestHandler(RelayedHttpListenerContext context, Strin
 	}
 	
 	private static void httpRequestSender(String method, String msgExpected, String msgToSend) throws IOException {
+		HttpURLConnection connection = getHttpConnection();
+		sendBytesOverHttp(connection, msgToSend.getBytes(), 0, msgToSend.length());
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >msgToSend.length() [](start = 57, length = 18)

call msgToSend.getBytes() (UTF-8?), hang on to the byte[] and use its length here instead of String.length() (the UTF-16 character count).

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 9]
```diff
@@ -280,29 +339,36 @@ private static void httpRequestHandler(RelayedHttpListenerContext context, Strin
...
-		conn.setDoOutput(true);
-		OutputStreamWriter out = new OutputStreamWriter(conn.getOutputStream());
-		out.write(message, 0, message.length());
+		HttpURLConnection connection = (HttpURLConnection)url.openConnection();
+		connection.setRequestProperty("ServiceBusAuthorization", tokenString);
+		return connection;
+	}
+	
+	private static void sendBytesOverHttp(HttpURLConnection connection, byte[] b, int off, int len) throws IOException {
+		connection.setRequestMethod((len > 0) ? "POST" : "GET");
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >"GET" [](start = 51, length = 5)

For future consideration: Do the following verbs work in our Java Library?

Delete, Get, Head, Options, Post, Put, Trace, Patch

Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >( [](start = 30, length = 1)

total style nit but consider omitting these parenthesis if they're not needed.

Comment (dlstucki) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: I said this then I started to feel a bit silly. (At this point you have more Java experience on the team than I do and perhaps the Java conventions should come from you). Feel free to ignore. We do have a .NET StyleCop rule which enforces this in C# though. 

---
In reply to: [294385721](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294385721) [](ancestors = 294385721)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 10]
```diff
@@ -326,18 +392,28 @@ private static void sendResponseMessages(RelayedHttpListenerContext context, Str
 		}
 	}
 	
-	private static String readStringFromStream(InputStream inputStream) throws IOException {
+	private static byte[] readBytesFromStream(InputStream inputStream) throws IOException {
 		if (inputStream == null) {
-			return emptyStr;
+			return new byte[0];
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: > new byte[0]; [](start = 9, length = 13)

Food for thought: I'm curious if this has the result of converting a null inputStream, kind of like (String)null, into a "". Does this make it harder to test any variations we might be interested in?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 2 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: Do you mean replace the current tests using small/large strings with these bytes repeated? I'm not sure what kind of difference it would make with different sizes like 1024 vs 4096

---
In reply to: [294377702](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294377702) [](ancestors = 294377702,294017638)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 4 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: HCClient has no close(), closing the connection it creates instead.

---
In reply to: [294380700](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294380700) [](ancestors = 294380700)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 5 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: Do you feel like there should be separate tests that just focus on the status code and description? I didn't include any because I didn't think there are any special values that we need to test specifically

---
In reply to: [294381396](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294381396) [](ancestors = 294381396)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 2 above)
Comment (dlstucki) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: First of all I'm trying to get you outside of testing strings. The library deals with bytes not Strings as near AFAICT. Strings convert to bytes in a MANY different ways (which are outside the scope of the library here).

I want to ensure if the library is given these 11 bytes:
0xFF 0xFE 0x65 0x00 0x66 0x00 0x67 0x00 0x00 0x00 0x00 (UTF-16LE with byte order mark string of "ABC" plus extra 0x00 0x00 terminator plus one extra 0x00 byte to make it an odd number)

that our library doesn't convert it to something else like these 4 bytes which once converted to String are equivalent to the original 10 bytes:
0x65 0x66 0x67 0x00            (ASCII or UTF-8 with no byte order mark, lost trailing 0x00 bytes).

---
In reply to: [295527790](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295527790) [](ancestors = 295527790,294377702,294017638)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 5 above)
Comment (dlstucki) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: I'd feel better the tests do at least 1 variation (2 total) on StatusCode. Otherwise how do you know it's actually using the one you told it to? Same thing with Description.

---
In reply to: [295535676](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295535676) [](ancestors = 295535676,294381396)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 10 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: I did this to avoid potentially having null checks everywhere.

The only side effect I can think of is if we send and receive null vs. byte[0], but according to the code, sending null would just return and do nothing, so I don't think we need to worry about tests sending and receiving null vs. byte[0].

---
In reply to: [294387162](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294387162) [](ancestors = 294387162)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 9 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: Will include in future PR

---
In reply to: [294385150](https://github.com/Azure/azure-relay-java/pull/47#discussion_r294385150) [](ancestors = 294385150)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 11]
```diff
@@ -106,19 +103,19 @@ public void websocketRepeatedSendReceiveTest() {
 		CompletableFuture<Integer> senderReceiveCount = new CompletableFuture<Integer>();
 		CompletableFuture<Integer> listenerReceiveCount = new CompletableFuture<Integer>();
 		
-		listener.acceptConnectionAsync().thenAcceptAsync((websocket) -> {
+		listener.acceptConnectionAsync().thenAcceptAsync(websocket -> {
 			for (int i = 1; i <= timesToRepeat; i++) {
 				websocket.readAsync().join();
-				websocket.writeAsync(StringUtil.toBuffer("hi")).join();
+				websocket.writeAsync(ByteBuffer.wrap(SMALL_BYTES)).join();
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >ByteBuffer.wrap(SMALL_BYTES)). [](start = 25, length = 30)

What impact does this comment from ClientWebSocket.java have on the shared SMALL_BYTES array?
"sendBinary() will cause the content of the byte array within the ByteBuffer to change"

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 12]
```diff
@@ -172,147 +169,148 @@ public void websocketMultipleSenderTest() throws URISyntaxException {
...
 		});
 	}
 	
-	private static CompletableFuture<Void> websocketListener(String msgExpected, String msgToSend) {
-		return listener.acceptConnectionAsync().thenComposeAsync((channel) -> {
+	private static CompletableFuture<Void> sendAndReceiveWithWebsocketListener(byte[] msgExpected, byte[] msgToSend) {
+		return listener.acceptConnectionAsync().thenComposeAsync(channel -> {
 			return channel.readAsync().thenAccept(bytesReceived -> {
-				assertEquals("Websocket listener did not receive the expected message.", msgExpected, new String(bytesReceived.array()));
+				assertTrue("Websocket listener did not receive the expected message.", Arrays.equals(msgExpected, bytesReceived.array()));
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >bytesReceived.array() [](start = 102, length = 21)

If the ByteBuffer ever changes to not happen to give you a perfectly sized array this will fail, right?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 11 above)
Comment (dlstucki) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: is that why some places call .clone()?

---
In reply to: [295925633](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295925633) [](ancestors = 295925633)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 5 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: That makes sense, I will try to take that into account probably in a future PR.

---
In reply to: [295557975](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295557975) [](ancestors = 295557975,295535676,294381396)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 11 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: Yes, `.clone()` just to avoid the content getting messed up

---
In reply to: [295927864](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295927864) [](ancestors = 295927864,295925633)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java [hunk 13]
```diff
@@ -106,19 +103,19 @@ public void websocketRepeatedSendReceiveTest() {
...
 					listenerReceiveCount.complete(i);
 				}
 			}
 		});
 		
-		client.createConnectionAsync().thenAccept((clientWebSocket) -> {
+		client.createConnectionAsync().thenAccept(clientWebSocket -> {
 			for (int i = 1; i <= timesToRepeat; i++) {
-				clientWebSocket.writeAsync(StringUtil.toBuffer("hi")).join();
+				clientWebSocket.writeAsync(ByteBuffer.wrap(SMALL_BYTES)).join();
```
Comment (dlstucki) [inline src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: >SMALL_BYTES [](start = 47, length = 11)

clone here?

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 11 above)
Comment (dlstucki) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: Let's put clone into our ClientWebSocket then.

---
In reply to: [295930809](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295930809) [](ancestors = 295930809,295927864,295925633)

File: src/test/java/com/microsoft/azure/relay/SendReceiveTest.java (hunk 12 above)
Comment (bainian12345) [reply src/test/java/com/microsoft/azure/relay/SendReceiveTest.java]: It would fail, but currently `readAsync` returns a perfectly wrapped ByteBuffer. It's not like we are letting the user bring their own buffer anyways as of now

---
In reply to: [295927555](https://github.com/Azure/azure-relay-java/pull/47#discussion_r295927555) [](ancestors = 295927555)
